lib/util/DateTimeHelpers.ts

```diff
@@ -0,0 +1,195 @@
+  const time = timeStrings.slice(0, -1).map(str => Number(str));
+
+  const res = {
+    hours: time[0],
+    minutes: time[1],
+    seconds: time[2],
+    ...timeZoneParser(timeStrings[3]),
+  };
+
+  if (res.seconds >= 60 || res.minutes >= 60 || res.hours > 24 ||
+    (res.hours === 24 && (res.minutes !== 0 || res.seconds !== 0))) {
+    throw new Error('nono');
+  }
+  return res;
+}
+
+export function timeParser(timeStr: string, errorCreator?: () => Error): ITimeRepresentation {
+  // https://www.w3.org/TR/xmlschema-2/#time-lexical-repr
+  const res = internalTimeParser(timeStr, errorCreator);
+  res.hours %= 24;
+  return res;
+}
+
+export function timeSerializer(time: ITimeRepresentation): string {
+  return `${numSer(time.hours)}:${numSer(time.minutes)}:${numSer(time.seconds)}${timeZoneSerializer(time)}`;
+}
+
+export function durationParser(durationStr: string): Partial<IDurationRepresentation> {
+  // https://www.w3.org/TR/xmlschema-2/#duration-lexical-repr
+  const [ dayNotation, timeNotation ] = durationStr.split('T');
+
+  // Handle date part
+  const formattedDayDur = dayNotation.replace(/^(-)?P(\d+Y)?(\d+M)?(\d+D)?$/gu, '$11S!$2!$3!$4');
+  if (formattedDayDur === dayNotation) {
+    throw new Error('nono');
+  }
+
+  const durationStrings = formattedDayDur.split('!');
+  if (timeNotation !== undefined) {
+    const formattedTimeDur = timeNotation.replace(/^(\d+H)?(\d+M)?(\d+(\.\d+)?S)?$/gu, '$1!$2!$3');
+
+    if (timeNotation === '' || timeNotation === formattedTimeDur) {
+      throw new Error('nono');
+    }
+    durationStrings.push(...formattedTimeDur.split('!'));
+  }
+  const duration = durationStrings.map(str => str.slice(0, -1));
+  if (!duration.slice(1).some(item => item)) {
+    throw new Error('nono');
+  }
+
+  const sign = <-1 | 1> Number(duration[0]);
+  return {
+    year: duration[1] ? sign * Number(duration[1]) : undefined,
+    month: duration[2] ? sign * Number(duration[2]) : undefined,
+    day: duration[3] ? sign * Number(duration[3]) : undefined,
+    hours: duration[4] ? sign * Number(duration[4]) : undefined,
+    minutes: duration[5] ? sign * Number(duration[5]) : undefined,
+    seconds: duration[6] ? sign * Number(duration[6]) : undefined,
+  };
+}
+
+export function yearMonthDurationParser(durationStr: string): Partial<IYearMonthDurationRepresentation> {
+  const res = durationParser(durationStr);
+  if (Object.entries(res).some(([ key, value ]) => value !== undefined && ![ 'year', 'month' ].includes(key))) {
+    throw new Error('nono');
+  }
+  return res;
+}
+
+export function dayTimeDurationParser(durationStr: string): Partial<IDayTimeDurationRepresentation> {
+  const res = durationParser(durationStr);
+  const filtered = extractDayTimeDur(res);
+  if (Object.entries(res).some(([ key, value ]) => value !== undefined && ![
+    'hours',
+    'minutes',
+    'seconds',
+    'day',
+  ].includes(key))) {
+    throw new Error('nono');
+  }
+  return res;
+}
+
```

Comment:
I am wondering if I could do this in a better way. @rubensworks Do you have a suggestion? (Don't mind the error messages)

Reply:
This might be a bit more compact, and slightly more performant: 

`if([ 'year', 'month' ].some(key => !(key in res)))`

But this assumes that keys in `res` always have a defined value.

(if this approach works, I would suggest moving this constant array outside of this function as an actual constant, to avoid re-recreating it for every call.

Reply:
They do not always have a defined value. My biggest issue is actually that both `year` and `month` in the first function are the 2 keys in `IYearMonthDurationRepresentation`.  So this code gets some typescript keys into JS runtime, but the problem is ofcource that the compiler will not catch the renaming of the `year` field into `years` for example. So basically I would want to get a list of all the keys for the `IYearMonthDurationRepresentation` interface. 
As in this issue: https://stackoverflow.com/questions/43909566/get-keys-of-a-typescript-interface-as-array-of-strings

Reply:
Yeah, that's not possible indeed, as TS typings are not available at runtime.

But if `undefined` is possible, then the following may work:

`if([ 'year', 'month' ].some(key => !res[key]))`

Reply:
Alright, that's unfortunate. I'll make sure to add some comments than :)
Your code fragment also looks better :) Thank you!